### ISSUE TITLE
black square box while generating pdf (unicode error)

### ISSUE THREAD

**Issue by beebek** [xhtml2pdf/xhtml2pdf#129]
A weird problem. While generating pdf, inplace of unicodes square black boxes apperars. Dont know if its unicode or font-face error. I even dont know if to use the "font-face and font-family" to generate the unicode into pdf. Anything I am missing ??
Great thanks.

Code snippet
    # -*- coding: utf-8 -*-
     
    from xhtml2pdf import pisa
    from StringIO import StringIO
     
    source = """<html>
                <style>
                    @font-face {
                    font-family: Mangal;
                    src: url("mangal.ttf");
                    }
                   
                    body {
                    font-family: Mangal;
                    }
                </style>
                <body>
                    This is a test <br/>
                           सरल
                </body>
            </html>"""
     
    # Utility function
    def convertHtmlToPdf(source):       
        pdf = StringIO()
        pisaStatus = pisa.CreatePDF(StringIO(source.encode('utf-8')), pdf)
     
        # return True on success and False on errors
        print "Success: ", pisaStatus.err
        return pdf
     
    # Main program
    if __name__=="__main__":
        print pisa.showLogging()
        pdf = convertHtmlToPdf(source)
        fd = open("test.pdf", "w+b")
        fd.write(pdf.getvalue())
        fd.close()

**Comment by teeberg** [comment]
All this has been mentioned in this thread before in some form, but it took me still a while to get it to run. The important parts for me were:
```css
@font-face {
    font-family: "FONTNAME";
    src: url('/absolute/path/to/yourfonts/somefontfile.ttf')
}
```
```css
body {
    font-family: "FONTNAME";
}
```
If you still get black/white boxes, make sure the font you use actually defines glyphs for the unicode code points that you are using! If looking at the HTML in your browser seem to works well but not using xhtml2pdf, it's probably because your browser falls back to other fonts (e.g. Times New Roman).

**Comment by benjaoming** [comment]
Closing this as a duplicate of #11 which should be fixed before we can release 0.1.

However lots of good info, props to everyone for sharing solutions!

**Comment by serboctor** [comment]
This was never fixed, right?
I still explicitly need set the font-family for the body if I want non-latin characters to appear?

**Comment by benjaoming** [comment]
@serboctor you're commenting on a closed ticket, please see #11

**Comment by SimonAstani** [comment]
> Welcome...
> If you want any kind of python support or related to flask do remember me.
> We do outsourcing as well. So if you need any kind of works to be done, we
> are here for you.
> 
> On Fri, Sep 6, 2013 at 3:48 PM, Lazarev Eugene [notifications@github.comwrote](mailto:notifications@github.comwrote):
> 
> > it's worked!!!! awesome!!!
> > thanks again, and if you need any programming help we can provide it to you
> > trouble was in wraping src in StringIO.StringIO() that doesn't need
> > before
> > pdf = pisa.pisaDocument(StringIO.StringIO(html), result)
> > after
> > pdf = pisa.pisaDocument(html, result)
> > —
> > Reply to this email directly or view it on GitHubhttps://github.com/[/issues/129](https://github.com/xhtml2pdf/xhtml2pdf/issues/129)#issuecomment-23930265
> > .
> 
> Regards,
> Bibek Chitrakar
Hello bibek, Your solution seems helpful. After providing relative path for the font. Only one fonts type get render. What if I want to render tow types of fonts instead at once..

**Comment by sunnybear** [comment]
It's very disappoiting to add local fonts for any HTML document to render in non-latin encoding
